src/internet_identity/src/state/temp_keys.rs

```diff
@@ -0,0 +1,150 @@
+use crate::MINUTE_NS;
+use candid::Principal;
+use ic_cdk::api::time;
+use internet_identity_interface::internet_identity::types::{AnchorNumber, DeviceKey, Timestamp};
+use std::cmp::Ordering;
+use std::collections::{BinaryHeap, HashMap};
+
+// Expiration for temp keys, the same as the front-end delegation expiry
+const TEMP_KEY_EXPIRATION_NS: u64 = 10 * MINUTE_NS;
+
+#[derive(Default, Debug)]
+pub struct TempKeys {
+    /// A map of "temporary keys" attached to devices (and a specific anchor). A temporary key can be used in lieu
+    /// of the device but has a short expiration time. These keys are used as a workaround for WebAuthn
+    /// needing two users interactions: one for "create" and one for "sign". So instead we only "create"
+    /// and instead authenticate the user with a temporary key for their first visit.
+    ///
+    /// Note: we link the temporary keys to a device so that we can make sure the temporary key is dropped
+    /// if the device itself is removed. This ensures the temporary key is no more powerful than the device,
+    /// i.e. if the user decides to remove the device right after registration, then the the temporary key
+    /// cannot be used to authenticate (similarly to how the browser session key pair cannot be used to
+    /// authenticate if the actual delegated WebAuthn device is removed).
+    ///
+    /// In addition, the temporary key is also linked to an anchor so that even if the same device is
+    /// added to multiple anchors, it can only be used in the context of the anchor it was created for.
+    ///
+    /// Since temp keys can only be added during registration, the the max number of temp keys is
+    /// bounded by the registration rate limit.
+    temp_keys: HashMap<DeviceKey, TempKey>,
+
+    /// Heap to efficiently prune expired temp keys
+    expirations: BinaryHeap<TempKeyExpiration>,
```

Comment:
Have you had performance issues with iterating the HashMap? How many temp keys do we expect at any given time?

Out of curiosity: why use a binary heap? Wouldn't a double ended queue be more efficient with O(1) push and pop? Or do you not expect to insert `TempKeyExpiration` "monotonously"?